src/lib/reasoners/uf.ml

```diff
@@ -1099,15 +1035,193 @@ module MED = Map.Make
+      acc
+    else
+      begin
+        let arg_vals, arg_tys, mrepr =
+          List.fold_left
+            (fun (arg_vals, arg_tys, mrepr) arg ->
+               let rep_arg, mrepr = model_repr_of_term arg env mrepr in
+               rep_arg :: arg_vals,
+               (Expr.type_info arg) :: arg_tys,
+               mrepr
+            )
+            ([], [], mrepr) (List.rev xs)
+        in
+        let ret_rep, mrepr = model_repr_of_term t env mrepr in
+        match f, arg_vals, ty with
+        | Sy.Name _, [], Ty.Tfarray _ ->
+          begin
+            match Hashtbl.find_opt cache.array_selects t with
+            | Some _ -> acc
+            | None ->
+              (* We have to add an abstract array in case there is no
+                 constraint on its values. *)
+              Hashtbl.add cache.array_selects t (Hashtbl.create 17);
+              acc
+          end
+
+        | Sy.Op Sy.Set, _, _ -> acc
+
+        | Sy.Op Sy.Get, [a; i], _ ->
+          begin
+            store_array_select a i ret_rep;
+            acc
+          end
+
+        | Sy.Name { hs = id; _ }, _, _ ->
+          let value =
+            match ty with
+            | Ty.Text _ ->
+              (* We cannot produce a concrete value as the type is abstract.
+                 In this case, we produce an abstract value with the appropriate
+                 type. *)
+              get_abstract_for env t
+            | _ -> ret_rep
+          in
+          ModelMap.(add (id, arg_tys, ty) arg_vals value mdl), mrepr
+
+        | _ ->
+          Printer.print_err
+            "Expect a uninterpreted term declared by the user, got %a"
+            E.print t;
+          assert false
+      end
+
+let extract_concrete_model cache =
+  let compute_concrete_model_of_val = compute_concrete_model_of_val cache in
+  let get_abstract_for = Cache.get_abstract_for cache.abstracts
+  in fun ~prop_model env ->
+    let terms, suspicious = terms env in
+    let model, mrepr =
+      MED.fold (fun t _mk acc -> compute_concrete_model_of_val env t acc)
+        terms (ModelMap.empty ~suspicious, ME.empty)
+    in
+    let model =
+      Hashtbl.fold (fun t vals mdl ->
+          (* We produce a fresh identifiant for abstract value in order to
+             prevent any capture. *)
+          let abstract = get_abstract_for env t in
+          let ty = Expr.type_info t in
+          let arr_val =
+            Hashtbl.fold (fun i v arr_val ->
+                Expr.ArraysEx.store arr_val i v
+              ) vals abstract
+          in
+          let id =
+            let Expr.{ f; _ } = Expr.term_view t in
+            match f with
+            | Sy.Name { hs; _ } -> hs
+            | _ ->
+              (* We only store array declarations as keys in the cache
+                 [array_selects]. *)
+              assert false
+          in
+          let mdl =
+            if not @@ Id.Namespace.Internal.is_id (Hstring.view id) then
```

Comment:
Would checking the `defined` field in the `Sy.Name` make more sense here?

Reply:
It doesn't. The `defined` field means the user defined the semantic of the symbols in the input file with `define-fun`. Here we're looking for array identifiers that haven´t been declared by the user. For instance in the input:
```smt2
(set-option :produce-models true)
(set-logic ALL)
(declare-sort S 0)
(declare-const s S)
(declare-datatype Pair ((pair (first (Array Int S)) (second (Array Int S)))))
(declare-const x Pair)
(assert (= (select (first x) 0) s))
(assert (= (first x) (second x)))
(check-sat)
(get-model)
```
We'll produce a identifier for embedded array in `x` but this array isn't declared by the user. Thus we don't want to display the definition of this array at the top level of the model.

Reply:
Right — we'd need `declared`, not `defined`, but we don't have that, of course. Thanks for the explanation